File: 프로그래머스/1/468370. 중요한 단어를 스포 방지/중요한 단어를 스포 방지.java
import java.util.*;

class Solution {
    public int solution(String message, int[][] spoiler_ranges) {
        
        Set<String> normal = new HashSet<>();
        Set<String> spoiler = new HashSet<>();
        
        int idx = 0;
        
        for (String word : message.split(" ")) {
            int start = idx;
            int end = start + word.length() - 1;
            idx = end + 2;
            
            boolean isSpoiler = false;
            
            for (int[] range : spoiler_ranges) {
                if (start <= range[1] && range[0] <= end) {
                    isSpoiler = true;
                    break;
                }
            }
            
            if (isSpoiler) {
                spoiler.add(word);
            } else {
                normal.add(word);
            }
        }   
        
        spoiler.removeAll(normal);
        return spoiler.size();
    }
}

File: 프로그래머스/1/468370. 중요한 단어를 스포 방지/중요한 단어를 스포 방지.py
def solution(message, spoiler_ranges):
    
    spoiler_words = set()
    normal_words = set()
    
    idx = 0
    
    if "  " in message:
        raise RuntimeError("연속된 공백 발견")
    
    for word in message.split(" "):
        start = idx
        end = start + len(word) - 1
        idx = end + 2
        
        is_spoiler = any(
            start <= right and end >= left
            for left, right in spoiler_ranges
    )
                
        if is_spoiler:
            spoiler_words.add(word)
        else:
            normal_words.add(word)
    
    return len(spoiler_words - normal_words)